### ISSUE TITLE
pointless new instances being created

### ISSUE THREAD

**Issue by bluedoo** [npgall/cqengine#150]
@npgall I put a print statement in my constructor and noticed that whenever the below code runs, new instances are created...

What is the purpose of creating so many new instances if all I need to do is save the object? If a new instance is absolutely necessary then just create one, surely?

```
public Foo() {
  System.out.println("new foo created");
}
```

```
public final IndexedCollection<Foo> getFooCollection() {
        IndexedCollection<Foo> foos = new ConcurrentIndexedCollection<>(DiskPersistence.onPrimaryKeyInFile(Foo.ID_INDEX, new File("foos.dat")));
        foos.addIndex(NavigableIndex.onAttribute(Foo.ID_INDEX));
        return foos;
}
```

```
if (getFooCollection().add(foo)) {
    return true;
} else {
    return false;
}
```

### Output

```
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
new foo created
...
```

**Comment by bluedoo** [comment]
@npgall run it the first time, just one is printed, run it again more are printed, run it again more and more...

#### App.java

```

import com.googlecode.cqengine.ConcurrentIndexedCollection;
import com.googlecode.cqengine.IndexedCollection;
import com.googlecode.cqengine.index.navigable.NavigableIndex;
import com.googlecode.cqengine.persistence.disk.DiskPersistence;
import java.io.File;

public class App {

    public static final IndexedCollection<Foo> getFooCollection() {
        IndexedCollection<Foo> foos = new ConcurrentIndexedCollection<>(DiskPersistence.onPrimaryKeyInFile(Foo.ID_INDEX, new File("foos.dat")));
        foos.addIndex(NavigableIndex.onAttribute(Foo.ID_INDEX));
        return foos;
    }

    public static final Boolean create(final Foo foo) {
        try {
            return getFooCollection().add(foo);
        } catch (Exception ex) {
            return false;
        }
    }

    public static void main(String args[]) {
        create(new Foo());
    }
}

```

#### Foo.java

```

import com.googlecode.cqengine.attribute.SimpleAttribute;
import com.googlecode.cqengine.query.option.QueryOptions;
import java.util.UUID;

public class Foo {

    public final String id;

    public static final SimpleAttribute<Foo, String> ID_INDEX = new SimpleAttribute<Foo, String>() {
        @Override
        public String getValue(Foo foo, QueryOptions queryOptions) {
            return foo.id;
        }
    };

    public Foo() {
        System.out.println("new foo created");
        this.id = UUID.randomUUID().toString();
    }
}

```

**Comment by bluedoo** [comment]
@npgall This can be a major issue in some systems. For example in a system with unique id reservation...

Consider the following pseudocode:

```
public class Foo {
     public final ReservedId reservedId;

     public Foo() {
        this.reservedId = new ReservedId();
     }
}
```

```
public class ReservedId {
     public final String id;

     public ReservedId() {
        ReservedId reservedId = new ReservedId(generateRandomString());
        while (indexedCollection.contains(reservedId)) {
              reservedId = new ReservedId(generateRandomString());
        }
        indexedCollection.add(reservedId);
     }

    public ReservedId(final String id) {
         this.id = id;
    }
}
```

If a new instance is created without consent each time then an unused id will be added to the collection each time

**Comment by npgall** [comment]
***Re: assigning IDs in the constructor***
I disagree. I think that approach will lead to brittle code.

An id is usually a permanent identifier for some kind of record. For example a primary key for the object or row in a database.

If you generate a UUID and you assign this to the id field in the constructor, and you have marked the id field as final, how do you ever expect to load preexisting objects by id? 

How would you write unit tests to validate your business logic, if you had designed your domain objects in a way that prohibited instantiating them with preexisting ids?

Just because you can write your code this way, does not mean CQEngine should support it.

You can hack Kryo and CQEngine to enable serialization/deserialization of final fields which are assigned in the constructor, but you’ll find it much easier to avoid this pain, by just simplifying your domain objects or using simple POJOs instead. This advice is not unusual or specific to CQEngine. It applies to most cases where objects need to be instantiated by frameworks (CQEngine, Kryo, Hibernate, JAXB, Jackson etc.)

***Re: the issue of pointless objects being created.***
First of all, you should expect CQEngine to instantiate your objects at arbitrary times, if they are persisted on disk. For example it needs to deserialize/instantiate them to return them in ResultSets. It will even deserialize/instantiate some objects _speculatively_ (depending on the availability or lack of indexes) to evaluate them for a query; and so not all objects which are instantiated will actually be returned.

However when I ran your code, I did notice that more objects were being instantiated than expected. It turns out this is not an issue with CQEngine, it’s an issue with how you are generating/assigning UUIDs in the constructor of your Foo class.

When Kryo (as used by CQEngine) deserializes your Foo objects from disk, it assumes that it can safely invoke the no-arg constructor of your Foo class to instantiate them during the deserialization process. However Kryo then does not actually deserialize the original value into the id field, because your constructor has already assigned a new UUID to the field, and that field is marked final. This means the id of your objects is actually changing on-the-fly during each serialization-deserialization round trip. 

This in turn is causing your objects to be duplicated every time the collection is being reloaded from disk...

As you know, CQEngine allows multiple types of persistence - on-heap, off-heap or disk - to be used simultaneously with the same collection. The collection itself can be persisted on disk, while some indexes on the collection can be located in on-heap or off-heap memory (or other combinations thereof) etc.

To support these combinations, CQEngine automatically rebuilds the in-memory indexes (or disk indexes which are no longer present) when the collection itself is reloaded from disk and the application configures the collection to use those types of indexes again.

However, due to the way the id in your Foo objects is changing on-the fly when deserialized, this is causing duplicate objects to be added to every index including the primary key index - causing the number of objects in your collection to increase (perhaps double) every time the collection is reloaded.

Upon investigating this, I did realize though, that it would make sense to add a check to skip the index reinitialization for those indexes which are already on disk - so I will investigate that improvement. Currently those indexes are rescanned regardless. However in your case, you have a combination of disk persistence with on-heap indexes, so the rescanning/reinitialization is necessary.

In summary - basically this behavior is expected. These requirements on domain objects are for compatibility with serialization frameworks such as Kryo. As Kryo is one of the best known and best performing such libraries for Java, I don't plan to change it. If you can, you'll find it easier if you can adopt these conventions to make your objects more compatible.

Hope that helps,
Niall

**Comment by bluedoo** [comment]
@npgall Thank you for taking the time to investigate and reply so thoroughly to my issue, I'm glad we're able to improve on this amazing technology

> How would you write unit tests to validate your business logic, if you had designed your domain objects in a way that prohibited instantiating them with preexisting ids?

CQEngine should support immutable objects. How jackson does it is by providing annotations like `JSONCreator` and `JSONProperty` to indicate which constructor to use for init and which fields represent which properties respectively.

> When Kryo (as used by CQEngine) deserializes your Foo objects from disk, it assumes that it can safely invoke the no-arg constructor of your Foo class

The way I see it, this is a very dangerous assumption that should never be made because it's simply too broad. It's highly unlikely that every implementation will be able to sacrifice their no args constructor. Any time a developer makes their objects immutable they immediately forfeit the luxury of persistence? 

> However in your case, you have a combination of disk persistence with on-heap indexes, so the rescanning/reinitialization is necessary.

Would you recommend that I change from on-heap indexes to disk indexes in that case?

> As Kryo is one of the best known and best performing such libraries for Java

Is it really needed? Can't you just write `byte[]` to file and then read `byte[]` from file?

**Comment by npgall** [comment]
Sorry I didn't explain it well earlier. I'm not saying that immutability is bad. Actually I think the opposite to be true. I highly recommend that POJOs are immutable. In fact nearly all of the [example POJOs](https://github.com/npgall/cqengine/blob/master/code/src/test/java/com/googlecode/cqengine/examples/introduction/Car.java) in CQEngine documentation and the test suite are immutable.

CQEngine supports both mutable and immutable objects. It does not require a no-arg constructor specifically. However, if there is a no-arg constructor it will use it to instantiate the object, with the assumption that it can then assign values to the fields after instantiation because the presence of a no-arg constructor means the object must be _mutable_ and the fields will not be marked final.

This is because it typically makes no sense to have a no-arg constructor in a serializable immutable object. There would be no way to assign values to its fields in the first place, and there would be no way to modify its fields after it was instantiated either. So what purpose could it serve? This is the use case which CQEngine does not support.

If you have a mutable POJO, it's quite common that it may have a no-arg constructor, and it's also quite common that it might have other constructors as well, which allow subsets of the mutable fields to be populated at once, when the object is created. CQEngine supports both of these cases.

If you have an immutable POJO, it will typically NOT have a no-arg constructor, and it will typically only have multi-argument constructor(s) instead, which allows all of its fields to be populated. CQEngine supports this use case too.

**Now for some additional quirks**
I mentioned above that if the POJO has a no-arg constructor, that CQEngine will invoke it. However, note that if the POJO does not have a no-arg constructor, CQEngine will NOT actually invoke any constructor at all. For example, if the POJO only has multi-argument constructor, CQEngine will not invoke it. In this case CQEngine will use [Objenesis](http://objenesis.org/) to instantiate the object without calling any constructors.

Naturally, you might ask, why is this? The approach which invokes a no-arg constructor works on a wider variety of JVMs. The approach to bypass the constructor [does not work on some JVMs](https://github.com/easymock/objenesis/blob/master/SupportedJVMs.md), including on earlier versions of Android. Read about [Object creation](https://github.com/EsotericSoftware/kryo#object-creation) in Kryo literature for details.

So in summary, the developer does not sacrifice immutability. Immutability is supported and I recommend it. I don't really see a downside to the current design. Although nothing is perfect!

**Re: change from on-heap indexes to disk indexes** 
I would recommend: 
(1) use disk persistence with only disk indexes, or
(2) use on-heap persistence with on-heap indexes, or with disk indexes.
The reason for this is, if you have on-heap indexes they will as a side-effect retain copies of all the objects in the collection on-heap anyway, so this can defeat the purpose of using disk persistence to some degree.
However if you use on-heap persistence, it's ok to use disk indexes with this as there is no such side effect.

**Re: Can't you just write byte[] to file and then read byte[] from file**
How to convert the object to a byte[] in the first place? This is what Kryo does.
